Add unmet water heater loads EMS program

jmaguire1 opened the pull request:
## Pull Request Description
Add an EMS program for calculating unmet water heating loads. Focus is on unmet showers, but we can also calculate this for each end use if we wanted to. Addresses #1170.

## Checklist

PR Author: Check these when they're done. Not all may apply. ~~strikethrough~~ and check any that do not apply. 

PR Reviewer: Verify each has been completed.

- [ ] ~Schematron validator (`EPvalidator.xml`) has been updated~
- [x] Sample files have been added/updated (via `tasks.rb`)
- [x] Tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests` and/or `workflow/tests/hpxml_translator_test.rb`)
- [x] Documentation has been updated
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [ ] No unexpected changes to simulation results of sample files

jmaguire1 committed:
Add unmet loads EMS program

jmaguire1 committed:
busted by rubocop :)

joseph-robertson committed:
First cut at setting up annual unmet water heating loads.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

jmaguire1 committed:
Wrong direction for sag calculation

jmaguire1 committed:
Some updates to break out showers from "fixtures"

jmaguire1 committed:
One more change towards adding showers back in

jmaguire1 committed:
Add shower column to constants

jmaguire1 committed:
Fix shower schedules for smooth occupancy

jmaguire1 committed:
Still in progress, but: working on restructuring the calculation based on some discussions with Joe. Still need to connect the EMS sensor for WH outlet temperature.

jmaguire1 committed:
Working again. Still needs some cleanup and review

jmaguire1 committed:
Merge remote-tracking branch 'remotes/origin/master' into unmet_wh_loads

jmaguire1 commented:
Right now, this method calculates:

- Unmet shower energy (m*cp*delta T when T_delivered < 105 F)
- Unmet shower time
- Total shower time

Unmet shower energy I feel is the most complete metric, since it accounts for how far off from the desired temperature you are. However, no one really has a sense for if say 2 kWh of unmet showers is too many, so this way you can calculate the % of shower time where loads are unmet.

We could also add metrics related to ALL hot water usage, or all fixtures. We'd want to account for the different desired delivered temperature for appliances, but it wouldn't be hard to add metrics like Unmet Energy for all fixtures.

jmaguire1 committed:
Add to changelog, fix merge conflicts, rubocop

jmaguire1 committed:
Fix some typos

jmaguire1 committed:
rename constant for consistency with schedule:file

shorowit commented on HPXMLtoOpenStudio/measure.rb:
```diff
@@ -1317,6 +1319,8 @@ def self.add_hot_water_and_appliances(runner, model, weather, spaces)
     # Add combi-system EMS program with water use equipment information
     Waterheater.apply_combi_system_EMS(model, @hpxml.water_heating_systems, plantloop_map)
+
+    return plantloop_map[sys_id]
```
This currently just returns the last plant loop, which isn't what you want when there are multiple water heaters in a dwelling unit.

shorowit commented on HPXMLtoOpenStudio/resources/hotwater_appliances.rb:
```diff
@@ -964,6 +965,89 @@ def self.get_fixtures_gpd(eri_version, nbeds, fixtures_all_low_flow, daily_mw_fr
   end
 
+  def self.add_showers_and_calculate_max(model, runner, hpxml, weather, water_heating_system, eri_version, schedules_file, unavailable_periods)
+    nbeds = hpxml.building_construction.additional_properties.adjusted_number_of_bedrooms
+
+    # Process fixtures
+    fixtures_usage_multiplier = hpxml.water_heating.water_fixtures_usage_multiplier
+    fixtures_all_low_flow = true
+    hpxml.water_fixtures.each do |water_fixture|
+      next unless [HPXML::WaterFixtureTypeShowerhead, HPXML::WaterFixtureTypeFaucet].include? water_fixture.water_fixture_type
+
+      fixtures_all_low_flow = false if not water_fixture.low_flow
+    end
+
+    # Get distribution system
+    if hpxml.water_heating_systems.size > 0
+      hot_water_distribution = hpxml.hot_water_distributions[0]
+    end
+
+    # Calculate mixed water fractions
+    t_mix = 105.0 # F, Temperature of mixed water at fixtures
+    avg_setpoint_temp = 0.0 # WH Setpoint: Weighted average by fraction DHW load served
+    hpxml.water_heating_systems.each do |water_heating_system|
+      wh_setpoint = water_heating_system.temperature
+      wh_setpoint = Waterheater.get_default_hot_water_temperature(eri_version) if wh_setpoint.nil? # using detailed schedules
+      avg_setpoint_temp += wh_setpoint * water_heating_system.fraction_dhw_load_served
+    end
+    daily_wh_inlet_temperatures = calc_water_heater_daily_inlet_temperatures(weather, nbeds, hot_water_distribution, fixtures_all_low_flow, hpxml.header.sim_calendar_year)
+    daily_mw_fractions = calc_mixed_water_daily_fractions(daily_wh_inlet_temperatures, avg_setpoint_temp, t_mix)
```
This is all duplicated code from elsewhere. We should rework this code so that the new calculation & EMS program is simply created in the existing code as we process each water heater (rather than trying to do this all at the end of the measure).

jmaguire1 replied:
yeah, at one point I did have most of this integrated in with the fixtures, but I was having trouble figuring out to structure things. Let me take another shot at that and then if I get stuck I'll ask you or Joe for some help.

shorowit reviewed (commented):
Some comments while I think of it. Addressing these comments will make it easier to pull in the master branch.

jmaguire1 committed:
Major refactor of the unmet WH load calculation

jmaguire1 commented:
@shorowit: I just did a major refactor of this, no more duplicate code. Thanks for the suggestions. It'd be good to know if you think this is a decent way to organize things at a minimum if/when you have a chance to take a look, we're looking to use this with the 120 V HPWH relatively soon.

jmaguire1 committed:
Fix unmet energy calc (incorrectly passing shower peak flow rate)

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 committed:
remove print statement, rubocop

jmaguire1 committed:
Update some output variable names, add showers to unavailable_periods.csv

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

joseph-robertson committed:
Bit of code cleanup and adding FIXMEs to address.

joseph-robertson committed:
Oops, showers_schedule_obj does get used.

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads.

Attempt #1 at fixing all the conflicts

jmaguire1 committed:
oops, missed one update in the merge

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 committed:
Remove total shower time, only output unmet shower time (in hours) and unmet shower energy

jmaguire1 committed:
oops, missed removing shower time in one spot

jmaguire1 committed:
Some fixes for unmet showers with multiple WHs.

Still need to develop a few new test cases (since unmet loads is basically always expected to be 0 with smooth profiles, and that's especially true with multiple WHs) and do some testing for 2 tank SHWs (to exclude calculating unmet showers from the storage tank before the backup WH)

jmaguire1 committed:
Clean up EMS a bit, change hot water usage be 0 during outages (helps with testing, and technically more correct, even though there's no energy impact to draws when the WH is off)

jmaguire1 committed:
This should actually exclude solar storage tanks from 2 tank systems, but there's probably a more elegant way to do it.

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	HPXMLtoOpenStudio/resources/constants.rb
#	HPXMLtoOpenStudio/resources/hotwater_appliances.rb
#	HPXMLtoOpenStudio/resources/output.rb
#	HPXMLtoOpenStudio/resources/schedules.rb
#	ReportSimulationOutput/measure.rb

shorowit committed:
Bugfix.

shorowit committed:
tasks.rb update_hpxmls

shorowit committed:
Some more fixes/cleanup.

joseph-robertson committed:
Bunch of cleanup and test updates.

joseph-robertson committed:
Try a sample file with undersized tank water heater.

joseph-robertson committed:
Remove warning for both shower schedule types provided.

joseph-robertson committed:
Update reporting measure with include argument for unmet loads timeseries.

joseph-robertson committed:
Update reporting measure test for unmet loads outputs.

joseph-robertson committed:
Update run_simulation with unmet loads timeseries request option.

joseph-robertson committed:
Start updating the docs.

joseph-robertson committed:
Update the changelog entry.

joseph-robertson committed:
Few more fixes and updates.

joseph-robertson committed:
Fix comments. [ci skip]

joseph-robertson committed:
Catch target water temperature warnings for new dhw undersized sample file.

joseph-robertson committed:
Minor fix to ems program.

joseph-robertson committed:
Update the docs. [ci skip]

joseph-robertson committed:
Spaces in ems program and exceptions for 10x checks.

Commit:
Latest results.

shorowit commented on HPXMLtoOpenStudio/resources/waterheater.rb:
```diff
@@ -2262,11 +2262,11 @@ def self.unmet_wh_loads_program(model, water_heating_systems, plantloop_map, sho
-      unmet_wh_loads_program.addLine('Set ShowerSagTime=0')
-      unmet_wh_loads_program.addLine('Set ShowerE=0')
+      unmet_wh_loads_program.addLine('Set ShowerSagTime = 0.0')
+      unmet_wh_loads_program.addLine('Set ShowerE = 0.0')
```
Can we revert this change and fix the root cause? We should be able to handle the lack of spaces. It'd be nice to not have a developer hit this issue again in the future.

joseph-robertson committed:
Change energy annual units to kBtu and timeseries units to Btu.

joseph-robertson committed:
Relax expected results for dhw-undersized.

joseph-robertson committed:
Updates due to output name change.

Commit:
Latest results.

joseph-robertson committed:
Clean up loop that finds appropriate tank.

joseph-robertson committed:
Add sample files for undersized tankless, heat pump, and multiple.

joseph-robertson committed:
Refactor unmet loads program for multiple water heaters.

joseph-robertson committed:
Clean up expected warnings and checks.

joseph-robertson committed:
Typo.

joseph-robertson committed:
Expect err and run log warnings.

Commit:
Latest results.

jmaguire1 committed:
Change shower draws to be a % of fixtures (if smooth profiles) based on Scott's message.

Update unmet load program to work for any timestep and fix units issue.

jmaguire1 commented:
Thanks @joseph-robertson for helping me to get this working for multiple WHs!

I just finished some hand calculations and things look pretty great. There's some weirdness on the first timestep being counted as unmet, I might look into if there's anything I can do about that or not (I suspect not). I also want to do the hand calculations for multiple WHs and shorter timesteps to make sure that works as expected, if it does then this will be ready for review. Hopefully later this week.

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	Changelog.md
#	HPXMLtoOpenStudio/measure.xml
#	workflow/tests/base_results/results_simulations_bills.csv
#	workflow/tests/base_results/results_simulations_energy.csv
#	workflow/tests/base_results/results_simulations_hvac.csv
#	workflow/tests/base_results/results_simulations_loads.csv
#	workflow/tests/base_results/results_simulations_misc.csv

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Try WarmupFlag in program to avoid counting first timestep.

joseph-robertson commented:
> > There's some weirdness on the first timestep being counted as unmet,
> 
> Maybe related to the warmup period? Not too long ago, I addressed a similar issue where EC_adj was off for the first timestep, see [this commit](https://github.com/NREL/OpenStudio-HPXML/pull/1724/commits/eb50a8eeaa67ffc394ad6db88eaf8ebebc1faf66#diff-9316631e12103383267af4b98720e73e228d723d7a0faa552c5cf883e1b9e8b1).

Trying a similar approach in https://github.com/NREL/OpenStudio-HPXML/pull/1502/commits/593d488fcaba9ab5e336e2f703a9835ef70f6d9f.

joseph-robertson committed:
Add unmet loads method for building-level EMS program.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Skip building-level EMS program if no water heating systems.

joseph-robertson committed:
Try changing unit multiplier tolerances.

Commit:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

Commit:
Latest results.

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 committed:
Oops, messed this up in merging. Might still have to fix one or two more merge conflicts.

jmaguire1 committed:
Update unmet shower program to account for latest OS-HPXML changes

jmaguire1 committed:
Go back to more generic unmet loads program (rather than unmet shower loads program). We could potentially have other unmet loads in the future in this program, and outputs are already pretty explicitly labeled as shower related.

jmaguire1 committed:
missed one "." to "::" for stochastic showers

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 committed:
Fix schedules tests to match latest OS-HPXML updates

jmaguire1 committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

Commit:
Latest results.

jmaguire1 commented:
One thing I did realize after trying to use this in OCHRE: without having the shower max flow rate somewhere in HPXML or an unnormalized schedule, it's hard to do the equivalent comfort calculations. I could see shower max gpm being an HPXML extension or switching to an unnormalized schedule (unnormalized is probably easier), but I think I can address this in a follow up after I get at least some of the 120V HPWH capabilities in.

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	BuildResidentialScheduleFile/measure.xml
#	HPXMLtoOpenStudio/measure.rb
#	HPXMLtoOpenStudio/measure.xml
#	HPXMLtoOpenStudio/resources/data/unavailable_periods.csv
#	HPXMLtoOpenStudio/resources/hotwater_appliances.rb
#	workflow/tests/base_results/results_simulations_misc.csv

Commit:
Latest results.

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	HPXMLtoOpenStudio/measure.xml
#	workflow/tests/base_results/results_simulations_misc.csv
#	workflow/tests/util.rb

Commit:
Latest results.

shorowit commented on docs/source/workflow_outputs.rst:
```diff
@@ -325,6 +325,20 @@ Annual unmet hours are listed below.
 
 These numbers reflect the number of hours during the heating/cooling season when the conditioned space temperature deviates more than 0.2 deg-C (0.36 deg-F) from the heating/cooling setpoint.
 
+Annual Unmet Loads
+~~~~~~~~~~~~~~~~~~
+
+Annual unmet loads are listed below.
+
+  ========================================  =====
+  Type                                      Notes
+  ========================================  =====
+  Unmet Loads: Hot Water Shower Energy (J)  TODO
+  Unmet Loads: Hot Water Shower Time (hr)   TODO
+  ========================================  =====
+
+TODO
```
What if we combine these with the above Annual Unmet Hours like so:

- `Unmet Hours: Showers (hr)`
- `Unmet Hours: Showers (%)`

Also document that this accounts for a storage water heater running out of water and occupant discomfort. It does not account for temperature drop due to a distribution system or a tankless water heater's minimum firing rate. Using smooth schedules (rather than detailed/stochastic schedules) may under-represent these values.

joseph-robertson replied:
What should we report for timeseries, just `Unmet Hours: Showers (hr)`?

Also, seems like if we're no longer reporting unmet loads in kBtu, then we'd no longer need to use shower peak flow in the EMS program? If not, this should simplify down to just comparing Use Side temperature to Mixed temperature during shower draws.

shorowit replied:
Good point, yes the % output would only make sense as an annual output.

Second question is for @jmaguire1.

jmaguire1 replied:
Just the hours in the timeseries makes a lot of sense.

If we drop the energy metric, and just track hours of unmet loads, then we could drop the flow rate. The original intent with including flow rate and calculating energy, rather than just hours, was that we might want to account for difference between a shower that's just below setpoint and one that's like 30F below setpoint, or a small amount of flow vs. multiple showers at the same time being below setpoint. I've been proposing energy as the more important metric for a few years now, but the problem (as we discussed) is no one has any context of what 1 kWh of unmet showers means.

All that being said: I'm OK with dropping energy and trying to just report unmet hours of showering, in which case I think it's fine to drop max flow rate from the EMS program. There's no real standard way to report this, and there's advantages to hours in terms of being more intuitive.

joseph-robertson replied:
Sounds good @jmaguire1. I've gone ahead and pulled the peak flow stuff out of the EMS program.

shorowit commented on tasks.rb:
```diff
@@ -2084,6 +2085,12 @@ def apply_hpxml_modification_sample_files(hpxml_path, hpxml)
     end
+    if ['base-dhw-multiple-undersized.xml'].include? hpxml_file
+      hpxml_bldg.water_heating_systems.each do |water_heating_system|
+        water_heating_system.heating_capacity = 5000 if water_heating_system.water_heater_type == HPXML::WaterHeaterTypeStorage
+        water_heating_system.temperature = 100.0 if [HPXML::WaterHeaterTypeHeatPump, HPXML::WaterHeaterTypeTankless].include?(water_heating_system.water_heater_type)
```
This is kind of weird, the HPWH/tankless systems are not actually undersized, they just have their setpoint artificially lowered to result in unmet showers. At least for storage/HPWH, can't we just set a small tank volume instead? Maybe we don't need all the sample files. We can probably remove the tankless one, for example.

(Also, hopefully at some point in the future, we'll allow optional capacity inputs for HPWHs and tankless systems.)

joseph-robertson replied:
I tried instead reducing tank volume (e.g., to 1 gal). Water Heater Use Side Outlet Temperature doesn't appear to drop (even when using stochastic schedules).

shorowit replied:
Any idea, @jmaguire1?

jmaguire1 replied:
Definitely a little surprised, but I know we do have some code that scale the max flow rate to ensure that you never draw out more than the whole tank volume in an hour https://github.com/NREL/OpenStudio-HPXML/blob/e770f42307069ba1ed46467ddaee3866d2f70551/HPXMLtoOpenStudio/resources/waterheater.rb#L837

I think this means the water heater might be reducing the flow rate through it with stochastic schedules if you try to specify like 2 gal/min, the WH might only see lower flow rate. E+ will throw an error if you have a flow rate on the use side larger than the tank volume/hour, so we do need this code. If reducing the tank volume and heater capacity doesn't give you unmet loads, I'm not sure that there's a better way to do this without reducing the setpoint.

joseph-robertson replied:
Hmm, I'm definitely confused here. The code you are referencing above is in the `apply_solar_thermal` method. Looks like `storage_tank.setUseSideDesignFlowRate(UnitConversions.convert(storage_volume, 'gal', 'm^3') / 60.1)` is used only for stratified tanks, not mixed. I'm seeing that use side outlet temperature doesn't drop for an HPXML file with storage water heater (i.e., mixed).

joseph-robertson replied:
Here is 40 gal vs 1 gal:
![image](https://github.com/user-attachments/assets/84fae02b-6944-4ed2-8f64-0773398730e9)

shorowit replied:
> If reducing the tank volume and heater capacity doesn't give you unmet loads, I'm not sure that there's a better way to do this without reducing the setpoint.

I am confused as well. This new output seems meaningless if it doesn't capture unmet hours associated with not having a large enough storage tank. Isn't that exactly the kind of situation we're trying to evaluate in the context of a (lower heating capacity, potentially without backup electric resistance) HPWH?

joseph-robertson replied:
@jmaguire1 Any ideas here?

jmaguire1 replied:
So I did do some runs here looking at super small tank volumes. Here's some of the relevant results for an undersized tank (https://github.com/NREL/OpenStudio-HPXML/blob/unmet_wh_loads/workflow/sample_files/base-dhw-undersized.xml). Relevant WH info:

- 40 gal nominal volume, electric storage water heater
- 1100 Btu/hr burners
- Mixed tank model
- Stochastic draw profile

![image](https://github.com/user-attachments/assets/2a8d6911-07bd-469d-999b-da6950f35b16)

I think this is pretty much what we'd expect to see, that with the super small burner capacity, the tank temperature DOES drop low enough that you get unmet loads, and we see them being tracked when there's a shower draw and the outlet temp is < 105 F (~40.56 C)?

With just dropping the tank volume down, you can still have the WH meet the load for smaller draws. It basically acts like a tiny tankless WH (I think we actually do assume ~1 gal volume for tankless water heaters right now, but a huge burner capacity. See https://github.com/NREL/OpenStudio-HPXML/blob/32d0767c8ad44c548b3cfe6f5e5fc9553a420dea/HPXMLtoOpenStudio/resources/waterheater.rb#L106)

shorowit commented on Changelog.md:
```diff
@@ -34,6 +34,7 @@ __New Features__
 - **Breaking change**: Disaggregates "Walls" into "Above Grade Walls" and "Below Grade Walls" in results_design_load_details.csv output file.
 - Updates `openei_rates.zip` with the latest residential utility rates from the [OpenEI U.S. Utility Rate database](https://apps.openei.org/USURDB/).
 - Adds a warning if the sum of supply/return duct leakage to outside values is very high.
+- Adds metrics for unmet hot water loads.
```
Need a better description.

joseph-robertson replied:
https://github.com/NREL/OpenStudio-HPXML/pull/1502/commits/6df3ec6b5f3e095325726e0958801cae2233f1c1

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Updates to sample files and reported outputs.

shorowit commented on ReportSimulationOutput/measure.rb:
```diff
@@ -467,32 +454,21 @@ def energyPlusOutputRequests(runner, user_arguments)
+    @unmet_hours.each do |key, unmet_hour|
+      unmet_program = unmet_hours_program
+      if key.include?(UHT::HotWaterShower)
+        unmet_program = unmet_loads_program
```
Rename to unmet_showers_program?

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Clean up new methods and fill out TODOs.

joseph-robertson committed:
Get name of the schedule, not the object.

joseph-robertson committed:
Avoid divide by zero when no dhw.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Update reporting measure descriptions and tests.

joseph-robertson committed:
Revert stuff in workflow tests util file.

joseph-robertson committed:
Update the docs. [ci skip]

Commit:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Update the changelog.

Commit:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

GitHub-Action committed:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Create a second sample file with 1 gal tank and 10-min stochastic schedules.

GitHub-Action committed:
Latest results.

joseph-robertson committed:
Shorten new sample file timestep to 10 min.

joseph-robertson committed:
Merge branch 'unmet_wh_loads' of github.com:NREL/OpenStudio-HPXML into unmet_wh_loads

jmaguire1 commented:
I know one of the questions here was why we're not seeing unmet loads if we just drop the tank size to 1 gal, and I think I know why.

If we have a 1 gal WH but 4.5 kW elements, we basically have a small tankless water heater. Pretty much like the under sink units you see some manufacturers make, and the reason the "very small" UEF bin exists. So something like these: https://www.rheem.com/group/rheem-tankless-electric-water-heaters-professional-classic-series-tankless-electric-rtex-3-5-rtex-06-kw/

These units can support flow rates up ~ 1 gal/min, depending on mains. Here's an example from those spec sheets (the number is element capacity in kW):
![image](https://github.com/user-attachments/assets/a136a78f-3469-473e-ab92-4f50e3b0e55b)

Even with stochastic events but hourly timesteps, you might end up with low enough flow rates that the "tankless" water heater we use for testing can keep up with the loads. So I think it's not actually unexpected that we might not see unmet loads if you just reduce the tank volume, depending on what the draw profile looks like. From our "base-dhw-undersized", here's the fixture flow rate in gal/min:

![image](https://github.com/user-attachments/assets/ca86eead-f953-46c4-b807-550fd1a58dde)

So a test case with 1 gal but 4.5 kW elements might actually keep up with the shower draws, and we'd get 0 unmet loads. I think it's much better for our test cases to just use small elements/burners, rather than small tanks. But to make sure this works right, Joe helped make a new test file that has a 1 gal tank and 4.5 kW elements. With hourly timesteps/events, that still produces 0 unmet loads, but with 10 minute timesteps/events you do see a tiny bit of unmet loads (one hour). So we DO now have a test case for this if we want to make sure both situations will lead to unmet loads.

jmaguire1 commented:
@shorowit: the other big outstanding question here was splitting out SSB from fixtures, since this is sensitive to shower flow rates. I'd vote that we not do that for now, but we could consider it in the future. I'm actually curious what feedback we get about this feature, since there's no real standard way to quantify unmet WH loads, and if people agree with me that we really should just focus on showers.

joseph-robertson committed:
Regenerate 10-min stochastic occupancy csv file.

GitHub-Action committed:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

GitHub-Action committed:
Latest results.

jmaguire1 commented:
@shorowit: Are there any additional changes that need to be made here? I know this didn't make release, but I think this is largely ready to go? Or are there more changes required before we look at merging this in?

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	HPXMLtoOpenStudio/measure.xml
#	workflow/tests/base_results/results_simulations_misc.csv

shorowit committed:
Some misc cleanup.

GitHub-Action committed:
Latest results.

shorowit committed:
Some more cleanup and comments.

shorowit committed:
Merge branch 'unmet_wh_loads' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

shorowit committed:
Minor cleanup [ci skip]

shorowit commented on docs/source/workflow_outputs.rst:
```diff
@@ -319,11 +319,16 @@ Annual unmet hours are listed below.
-  Unmet Hours: Heating (hr)  Number of hours where the heating setpoint is not maintained.
+  Unmet Hours: Heating (hr)  Number of hours where the heating setpoint is not maintained. [#]_
   Unmet Hours: Cooling (hr)  Number of hours where the cooling setpoint is not maintained.
+  Unmet Hours: Showers (hr)  Number of hours where the shower water temperature is less than 105 deg-F. [#]_
```
I discussed with @jmaguire1; this should really say water heater tank outlet temperature, not shower water temperature. I will change this.

shorowit commented on HPXMLtoOpenStudio/resources/waterheater.rb:
```diff
@@ -2189,4 +2189,103 @@ def self.get_water_heater_solar_fraction(water_heating_system, hpxml_bldg)
+        key_name: tank.name
+      )
+
+      program.addLine("If (#{shower_flow_sensor.name} > 0) && (#{wh_temp_sensor.name} < #{mixed_setpoint_sensor.name})")
```
After discussing with @jmaguire1... Since we no longer calculate an unmet *load*, we now only care whether there is a shower for the given timestep. To prevent confusion about how the shower schedule is used (and prevent the need for error-checking if the shower schedule does not correctly jive with the overall fixtures schedule), I'm going to change the shower schedule to be a 0/1 schedule that simply represents the presence of a shower rather than the shower usage.

shorowit replied:
I'm second-guessing this. If the detailed schedule is hourly, you'd like to use a fraction for how much of the hour there were showers. Also related: https://github.com/NREL/OpenStudio-HPXML/issues/1890

shorowit replied:
I also wonder if the `ShowerUnmetTime` and `ShowerTime` variables below should be multiplied by the schedule value, where the schedule value represents how much of the timestep is for showers (stochastic) or the likelihood of showers for a given timestep (non-stochastic).

shorowit reviewed (commented):
This is looking good to me. I'm going to address a couple comments I'm leaving below, but it should probably be good to merge after that.

shorowit commented:
Need to compare outputs for:
- Hourly. stochastic schedule
- Sub-hourly. stochastic schedule
- Hourly. non-stochastic schedule
- Sub-hourly. non-stochastic schedule

The first two are currently giving very different answers (in hours), I think because of https://github.com/NREL/OpenStudio-HPXML/issues/1890.

shorowit commented on docs/source/workflow_outputs.rst:
```diff
@@ -319,11 +319,16 @@ Annual unmet hours are listed below.
+  .. [#] Unmet heating/cooling hours reflect the number of hours during the heating/cooling season when the conditioned space temperature deviates more than 0.2 deg-C (0.36 deg-F) from the heating/cooling setpoint.
+
+  .. [#] Unmet showers hours only account for a storage water heater running out of water and occupant discomfort; they do not account for temperature drop due to a distribution system or a tankless water heater's minimum firing rate.
+         Using simple hourly schedules (rather than detailed/stochastic sub-hourly schedules) may under-represent these values.
```
Need to review this once the results are making sense.

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	BuildResidentialScheduleFile/measure.xml
#	HPXMLtoOpenStudio/measure.xml
#	ReportSimulationOutput/measure.xml

shorowit committed:
Merge branch 'schedule_file_interpolation' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	HPXMLtoOpenStudio/measure.xml
#	workflow/tests/base_results/results_simulations_misc.csv

shorowit commented:
Note: I pulled #1896 into here and changed the base branch.

shorowit committed:
Clarify number of occupants in stochastic schedule csvs. Add a couple more examples.

GitHub-Action committed:
Latest results.

GitHub-Action committed:
Latest results.

rajeee committed:
Add showers minutes column

rajeee committed:
Add showers minutes column - fixes

shorowit committed:
Merge branch 'master' of https://github.com/NREL/OpenStudio-HPXML into unmet_wh_loads

# Conflicts:
#	BuildResidentialScheduleFile/measure.xml
#	HPXMLtoOpenStudio/measure.xml
#	tasks.rb
#	workflow/tests/base_results/results_simulations_misc.csv

shorowit committed:
Fix CI failures

GitHub-Action committed:
Latest results.

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Merge branch 'master' into unmet_wh_loads

joseph-robertson committed:
Update reporting measure for hot water shower constants.

joseph-robertson committed:
Regenerate hpxml files and update schedule_files README.

joseph-robertson committed:
Add test files for HPWH with nonzero unmet showers.

joseph-robertson committed:
Fix validation test.

joseph-robertson committed:
Typo in new HPWH undersized sample files.

GitHub-Action committed:
Latest results.

jmaguire1 committed:
Changes for minutes of shower use when doing unmet shower calcs. Still doing some testing.
